### DISCUSSION TITLE
Populate existing overrides previous query on joinedload

### DISCUSSION THREAD

**Discussion by mmicek** [sqlalchemy/sqlalchemy#11882] [Usage Questions]
### Describe the bug

Hi, i have problem with joinedload query with and_ clause. I have two entities:

``` python
   class Book(BaseEntity):
       pages: Mapped[list["Page"]] = relationship(
          secondary=pages_m2m, back_populates="books"
       )

   class Page(BaseEntity]):
       active: Mapped[bool]
```

Now i am adding two Page objects to database, one with active = True and one with active = False.
Connecting both to one Book object. So now in database there is one Book and two Page with different "active" value connected to this one Book.

Running this two queries:

``` python

    query_1 = select(Book).options(
        joinedload(Book.pages.and_(Page.active == True))
    )
    query_2 = select(Book).options(
        joinedload(Book.pages.and_(Page.active == False))
    )

    result_1 = list((await new_session.scalars(query_1)).unique().all())
    result_2 = list((await new_session.scalars(query_2)).unique().all())
```

Now in both results i have one Page with active = True. Following the documentation they say you need to use populate_existing=True to make select to database and fetch records.
But running new query with this settings:

``` python
    query_1 = select(Book).options(
        joinedload(Book.pages.and_(Page.active == True))
    )
    query_2 = select(Book).execution_options(populate_existing=True).options(
        joinedload(Book.pages.and_(Page.active == False))
    )

    result_1 = list((await new_session.scalars(query_1)).unique().all())
    result_2 = list((await new_session.scalars(query_2)).unique().all())
```

It returns in BOTH results one Page with active = False. 

Isn't it wierd caching behaviuur even if it is not bug? How to run both queries in one session and have proper results. One result with Page (active=True) and one result with Page(active=False).

Normally this should be correct result of both queries.

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/14/orm/queryguide.html#populate-existing

### SQLAlchemy Version in Use

sqlalchemy = {extras = ["asyncio"], version = "^2.0.31"}

### DBAPI (i.e. the database driver)

pyodbc = "^5.0.1"

### Database Vendor and Major Version

Postgresql 15, psycopg = {extras = ["binary", "pool"], version = "^3.2.1"}

### Python Version

3.11

### Operating system

Ubuntu

### To Reproduce

```python
   class Book(BaseEntity):
       pages: Mapped[list["Page"]] = relationship(
          secondary=pages_m2m, back_populates="books"
       )

   class Page(BaseEntity]):
       active: Mapped[bool]

    query_1 = select(Book).options(
        joinedload(Book.pages.and_(Page.active == True))
    )
    query_2 = select(Book).execution_options(populate_existing=True).options(
        joinedload(Book.pages.and_(Page.active == False))
    )

    result_1 = list((await new_session.scalars(query_1)).unique().all())
    result_2 = list((await new_session.scalars(query_2)).unique().all())
```

### Error

```
# Copy the complete stack trace and error message here, including SQL log output if applicable.
```

**Comment by CaselIT** [comment]
Hi,

> Isn't it wierd caching behaviuur even if it is not bug? How to run both queries in one session and have proper results. One result with Page (active=True) and one result with Page(active=False).

That's how the session works. It keeps an identity map, so the book that's returned by the two queries, it's exactly the same object (you can verify this by doing `assert result_1[0] is result_2[0]`.

Do run both queries in the same session you would need to either expunge or expire the objects after the first query or use two different sessions

**Reply by mmicek** [reply]
Thank you.

**Reply by BezBartek** [reply]
@CaselIT 
Why this is how it supposed to work? Could you provide rationale for it?

I feel that this prevents relying in code on SQL alchemy ORM entities. 

I might process something under a transaction, and inside that transaction, fetching objects under different conditions, and if I unexpectedly, in two separate code branches fetches the same record, properties of that object would be overwritten?
What if I can't split it into two separate connections, because I am doing all of it under DB lock. 

If you can't have two instances of the same entity (same ID), then you are always exposed to unexpected interruption into your entity, by code, that have nothing to do with code that you wrote, only by the fact, that it utilizes the same transaction. 

Thank you for response

**Reply by CaselIT** [reply]
> Why this is how it supposed to work? Could you provide rationale for it?

You can find documentation here: https://docs.sqlalchemy.org/en/20/orm/session_basics.html
In general this is how many [data mapper](https://en.wikipedia.org/wiki/Data_mapper_pattern) framework work.

> I might process something under a transaction, and inside that transaction, fetching objects under different conditions, and if I unexpectedly, in two separate code branches fetches the same record, properties of that object would be overwritten?
What if I can't split it into two separate connections, because I am doing all of it under DB lock.
> If you can't have two instances of the same entity (same ID), then you are always exposed to unexpected interruption into your entity, by code, that have nothing to do with code that you wrote, only by the fact, that it utilizes the same transaction.

How would you track what changes should go in the database if you have multiple instances of the same object at the same time? The row is only one. For example:
```py
obj1 = session.scalars(q1).one()
obj2 = session.scalars(q2).one()
obj1.a = 'foo'
obj2.b = 'bar'
session.commit()
```
What happens when you commit? do we set a=foo or b=bar? both? only the last one? 

I would suggest avoiding the use of loader conditions if you plan on running multiple queries on the same object.

An alternative that you may find more suitable with your use case may be write-only collection, so that instead you can get different list of object as you please from the relationship:  https://docs.sqlalchemy.org/en/20/orm/large_collections.htm The docs empathize large collections, but it will work with any size relationship

Overall the API behaviour when using multiple queries and eager loader condition is probably not the best, but it's really not feasible to change this part.

**Reply by CaselIT** [reply]
@zzzeek not sure if you have other suggestion for this usage pattern

**Reply by zzzeek** [reply]
I would maintain the two pages collections separately so there is no ambiguity, using two relationships `Book.active_pages` and `Book.inactive_pages`.

**Reply by BezBartek** [reply]
@CaselIT 
Thank you for response and clarification. I suspected that it is because of it. 
But, what I would expect to get in a case that you described, is to get an exception, **that I have ambiguous changes on instance at the same session**.
I think it should raise `AmbiguousChangesException` on a line where you invoke `obj2.b = 'bar'` or on `session.commit()`. 

I feel that current behavior is a problem. I faced it recently. 

We make on our project fixtures, to keep transaction LVL isolation on each test.
We bind session in the way that session.commit() invocations are save points, we mock the server to use the same session and, therefore we have a single session, for each test, that is used by the tested API client, and test code. Without that, we would need to clean DB for each test. Which would be a big problem at a huge codebase.  
Because of that, we might be in troubles. Cuz any test is a situation where I might have unexpected behavior, because in each test I am calling multiple queries for exactly the same object. Of course in most tests, it does not generate problems, but when it does, it is really trick, and each test is the possibility of that.

**Reply by zzzeek** [reply]
what's ambiguous exactly?    if I have an object `a`, and I do `a.foo = 'bar'` and later `a.foo = 'bat'`, what's ambiguous?

**Reply by CaselIT** [reply]
> what's ambiguous exactly? if I have an object a, and I do a.foo = 'bar' and later a.foo = 'bat', what's ambiguous?

my example was supposing an api where the session would give you two different instances connected to the same pk after two queries. in that case modifying both objects would cause issues.

> I feel that current behavior is a problem. I faced it recently.

I don't follow what's your issue is in the test. If the code you are testing expect unrelated session, then either create unrelated ones or close the session between invocation to reset it.
Overall how session works cannot really change without rewriting large parts of the orm.

**Reply by BezBartek** [reply]
@zzzeek 
```python
x = session.scalar(select(...))
y = session.scalar(select(...))
x.foo = 'bar'
y.foo = 'bat'
```
Currently they are exactly same objects, so you overwrite 'foo', and there is no ambiguity.  
But if they would not be same objects as they are currently, there would be a ambiguity, cuz it is not clear which one is the right value to save. 
As I understand from what @CaselIT said, this fact is one of reasons why the sqlalchemy behave like this. 

In the response to his question: "What happens when you commit? do we set a=foo or b=bar? both? only the last one?" 
I am answering - In the world where x and y are not the same objects, as they are currently, I would raise an AmbiguousChangesException.

**Reply by mmicek** [reply]
This also happens when i am adding related objects.

```
book.pages.append(Page(active=False))
book.pages.append(Page(active=True))
book.pages.commit()

select(Book).options(
     joinedload(Book.pages.and_(Page.active == True))
)
```
This select will return both pages (true and false), due to session remembers pages when they where added.
I know it is expected behaviuor, but
it's very confusing and problematic.

**Reply by zzzeek** [reply]
```py
select(Book).options(
     joinedload(Book.pages.and_(Page.active == True))
)
```

that's why you would use `populate_existing` here, which would overwrite the `Book.pages` collection with what's being loaded in the specific SQL statement.

Overall I would recommend against using WHERE criteria in a joinedload.   It's not intended for general use.  if you are looking to emit a SQL statement that includes a JOIN with particular criteria, use [.join() and receive the individual entities you are seeking separately](https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#selecting-multiple-orm-entities-simultaneously)

**Reply by mmicek** [reply]
Thats right, but if i want to make both queries active == True and active == False than for both result i will get results for active == False.
But sometimes i want to do some other loaded options like subqueryload, to be able to do e.g some aggregations.

**Reply by zzzeek** [reply]
not if you use a separate join and dont try to put criteria into "joinedload", which is not the same thing as "join".

you are trying to have a single object, with a single collection, that has two different things in it at different times.  As I said, this is not what you want.   your choices are:

1. use select().join() and get two separate objects instead - **easiest**
2. use two different relationships so that the two collections are explicitly kept separate - **also easiest**
3. keep track of which objects you expect in the collection for a particular `Session` (or use separate sessions) - **awkward**
4. **extremely advanced and somewhat experimental**:  use [identity_token](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#queryguide-identity-token) to split the two queries into two different sets of objects - **supports the exact pattern you are trying noting that this is not typically an idiomatic SQLAlchemy pattern, option 1 or 2 are**

**Reply by CaselIT** [reply]
it's very hard to understand the use case you are trying to solve.

To recap options here are:
- use two sessions. you can also create a new session against the same connection, keeping the same db transaction
- expunge/expire the objects between queries
- close the session between queries
- use two relationships
- select related objects directly using select.join
- use write only relationship
- use identity_token 
- use another library

Changing how the session works is unrealistic in this case sorry. If none of the options above are suitable and this is something you cannot live with another library is likely more suited to your use case

**Reply by zzzeek** [reply]
here's a demo for the last one

```py
from __future__ import annotations

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy import Table
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import joinedload
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass

pages_m2m = Table(
    "pages_m2m", Base.metadata,
    Column("book_id", ForeignKey("book.id"), primary_key=True),
    Column("page_id", ForeignKey("page.id"), primary_key=True)
)

class Book(Base):
    __tablename__ = 'book'
    id: Mapped[int] = mapped_column(primary_key=True)
    pages: Mapped[list["Page"]] = relationship(
        secondary=pages_m2m, back_populates="books"
       )

class Page(Base):
    __tablename__ = 'page'
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    active: Mapped[bool]
    books: Mapped[list["Book"]] = relationship(secondary=pages_m2m, back_populates="pages")

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(Book(pages=[
    Page(name="p1", active=False),
    Page(name="p2", active=True),
    Page(name="p3", active=True),
    Page(name="p4", active=False),
]))
s.commit()



query_1 = select(Book).execution_options(identity_token="active_true").options(
    joinedload(Book.pages.and_(Page.active == True))
)
query_2 = select(Book).execution_options(identity_token="active_false").options(
    joinedload(Book.pages.and_(Page.active == False))
)

result_1 = s.scalars(query_1).unique().all()
result_2 = s.scalars(query_2).unique().all()

# different identities
assert result_1[0] is not result_2[0]

# separate collections
assert {p.name for p in result_1[0].pages} == {"p2", "p3"}
assert {p.name for p in result_2[0].pages} == {"p1", "p4"}



```

**Reply by mmicek** [reply]
Thank you for the explanation. This is very good trick with this identity_token.

**Reply by BezBartek** [reply]
Thanks guys, identity_token and ability to split transaction into multiple sessions solves my issue too.
